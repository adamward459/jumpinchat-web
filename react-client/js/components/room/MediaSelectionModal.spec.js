/* global jest, window, it, beforeEach, describe */

import React from 'react';
import { shallow } from 'enzyme';
import MediaSelectionModal from './MediaSelectionModal.react';

describe('<MediaSelectionModal />', () => {
  let mediaSelectionModal;
  let props;
  beforeEach(() => {
    window.Janus = {
      attachMediaStream: jest.fn(),
    };

    window.ga = jest.fn();
    navigator.mediaDevices = {
      getDisplayMedia: jest.fn(() => Promise.resolve()),
    };

    props = {
      forcePtt: false,
      isGold: false,
      modal: {
        selectedDevices: {
          audio: 'foo',
          video: 'foo',
        },
        onSelectDevice: jest.fn(),
      },
      videoQuality: {
        id: 'foo',
        dimensions: {
          width: 1,
          height: 2,
        },
        frameRate: 69,
      },
      audioPtt: true,
    };

    mediaSelectionModal = new MediaSelectionModal(props);

    mediaSelectionModal.props = props;
  });

  describe('componentDidUpdate', () => {
    beforeEach(() => {
      mediaSelectionModal.publish = jest.fn();
    });

    it('should publish when audio device has changed', () => {
      const prevProps = {
        modal: {
          selectedDevices: {
            audio: 'bar',
            video: 'bar',
          },
        },
        audioPtt: true,
      };

      mediaSelectionModal.componentDidUpdate(prevProps);
      expect(mediaSelectionModal.publish)
        .toHaveBeenCalledWith(props.videoQuality, 'foo', 'foo', false);
    });
  });

  describe('closeModal', () => {
    it('should close the modal', () => {
      mediaSelectionModal.setMediaSelectionModal = jest.fn();
      mediaSelectionModal.closeModal();
      expect(mediaSelectionModal.setMediaSelectionModal).toHaveBeenCalledWith(false);
    });
  });


  describe('video', () => {
    beforeEach(() => {
      props = {
        ...props,
        modal: {
          ...props.modal,
          mediaType: 'video',
          deviceList: {
            audio: [],
            video: [
              {
                deviceId: '123',
                label: 'foo',
              },
              {
                deviceId: '321',
                label: 'bar',
              },
            ],
          },
        },
      };
    });

    it('should have a close button', () => {
      const wrapper = shallow(<MediaSelectionModal {...props} />);

      expect(wrapper.find('.modal__Button-close').length).toEqual(1);
    });

    describe('video sources', () => {
      it('should render available feeds', () => {
        const wrapper = shallow(<MediaSelectionModal {...props} />);

        // should have 4 as the screenshare source and
        // 'no video' options are always shown in the list
        expect(wrapper.find('MediaSource').length).toEqual(4);
      });

      it('should show error message if error is truthy', () => {
        props = { ...props, error: { message: 'foo' } };
        const wrapper = shallow(<MediaSelectionModal {...props} />);

        expect(wrapper.find('.modal__Error').length).toEqual(1);
        expect(wrapper.find('.modal__Error').text()).toEqual('foo');
      });

      it('should show an error if no available sources', () => {
        props = {
          ...props,
          modal: {
            ...props.modal,
            deviceList: {
              video: [],
            },
          },
        };
        const wrapper = shallow(<MediaSelectionModal {...props} />);

        expect(wrapper.find('.modal__Error').length).toEqual(1);
      });
    });

    describe('audio sources', () => {
      beforeEach(() => {
        props = {
          ...props,
          modal: {
            ...props.modal,
            mediaType: 'audio',
            deviceList: {
              audio: [
                {
                  deviceId: '123',
                  label: 'foo',
                },
                {
                  deviceId: '321',
                  label: 'bar',
                },
              ],
              video: [],
            },
          },
        };
      });

      it('should render available feeds', () => {
        const wrapper = shallow(<MediaSelectionModal {...props} />);

        expect(wrapper.find('MediaSource').length).toEqual(2);
      });

      it('should show error message if error is truthy', () => {
        props = { ...props, error: { message: 'foo' } };
        const wrapper = shallow(<MediaSelectionModal {...props} />);

        expect(wrapper.find('.modal__Error').length).toEqual(1);
        expect(wrapper.find('.modal__Error').text()).toEqual('foo');
      });

      it('should show an error if no available sources', () => {
        props = {
          ...props,
          modal: {
            ...props.modal,
            deviceList: {
              audio: [],
            },
          },
        };
        const wrapper = shallow(<MediaSelectionModal {...props} />);

        expect(wrapper.find('.modal__Error').length).toEqual(1);
      });

      it('should show PTT button if forcePtt is false', () => {
        const wrapper = shallow(<MediaSelectionModal {...props} />);

        expect(wrapper.getElement()).toMatchSnapshot();
      });

      it('should not show PTT button if forcePtt is true', () => {
        props.forcePtt = true;
        const wrapper = shallow(<MediaSelectionModal {...props} />);

        expect(wrapper.getElement()).toMatchSnapshot();
      });
    });
  });
});
