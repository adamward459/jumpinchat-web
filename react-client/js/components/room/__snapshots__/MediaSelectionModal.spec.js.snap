// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaSelectionModal /> video audio sources should not show PTT button if forcePtt is true 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal__Window"
  closeTimeoutMS={0}
  contentLabel="Media selection modal"
  isOpen={false}
  onRequestClose={[Function]}
  overlayClassName="modal"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <div
    className="modal__Header"
  >
    Choose 
    audio
     source
    <button
      className="modal__Button modal__Button-close"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-times"
      />
    </button>
  </div>
  <div
    className="modal__Body modal__Body--noPadding"
  >
    <React.Fragment>
      <React.Fragment>
        <ForwardRef
          className="scroll-shadow mediaSources__SourceListWrapper"
          contentClassName="mediaSources__SourceList"
          contentStyle={
            Object {
              "paddingBottom": "0.25em",
            }
          }
          horizontal={false}
        >
          <MediaSource
            device={
              Object {
                "deviceId": "123",
                "label": "foo",
              }
            }
            isGold={false}
            onSelectDevice={[Function]}
            type="audio"
            videoQuality={
              Object {
                "dimensions": Object {
                  "height": 2,
                  "width": 1,
                },
                "frameRate": 69,
                "id": "foo",
              }
            }
          />
          <MediaSource
            device={
              Object {
                "deviceId": "321",
                "label": "bar",
              }
            }
            isGold={false}
            onSelectDevice={[Function]}
            type="audio"
            videoQuality={
              Object {
                "dimensions": Object {
                  "height": 2,
                  "width": 1,
                },
                "frameRate": 69,
                "id": "foo",
              }
            }
          />
        </ForwardRef>
        <div
          className="mediaSources__ScreenShareContainer"
        >
          <Switch
            checked={true}
            disabled={true}
            helpText={null}
            id="mediaPttCheckbox"
            label="Push to talk"
            onChange={[Function]}
          />
          <span
            className="modal__SubText"
          >
            Push-to-talk is forced in this room
          </span>
        </div>
      </React.Fragment>
      <div
        className="modal__SubText"
      >
        Inappropriate broadcasts will be terminated as per the
         
        <a
          href="/terms"
          rel="noopener noreferrer"
          target="_blank"
        >
          terms and conditions
        </a>
      </div>
    </React.Fragment>
  </div>
</Modal>
`;

exports[`<MediaSelectionModal /> video audio sources should show PTT button if forcePtt is false 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal__Window"
  closeTimeoutMS={0}
  contentLabel="Media selection modal"
  isOpen={false}
  onRequestClose={[Function]}
  overlayClassName="modal"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <div
    className="modal__Header"
  >
    Choose 
    audio
     source
    <button
      className="modal__Button modal__Button-close"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-times"
      />
    </button>
  </div>
  <div
    className="modal__Body modal__Body--noPadding"
  >
    <React.Fragment>
      <React.Fragment>
        <ForwardRef
          className="scroll-shadow mediaSources__SourceListWrapper"
          contentClassName="mediaSources__SourceList"
          contentStyle={
            Object {
              "paddingBottom": "0.25em",
            }
          }
          horizontal={false}
        >
          <MediaSource
            device={
              Object {
                "deviceId": "123",
                "label": "foo",
              }
            }
            isGold={false}
            onSelectDevice={[Function]}
            type="audio"
            videoQuality={
              Object {
                "dimensions": Object {
                  "height": 2,
                  "width": 1,
                },
                "frameRate": 69,
                "id": "foo",
              }
            }
          />
          <MediaSource
            device={
              Object {
                "deviceId": "321",
                "label": "bar",
              }
            }
            isGold={false}
            onSelectDevice={[Function]}
            type="audio"
            videoQuality={
              Object {
                "dimensions": Object {
                  "height": 2,
                  "width": 1,
                },
                "frameRate": 69,
                "id": "foo",
              }
            }
          />
        </ForwardRef>
        <div
          className="mediaSources__ScreenShareContainer"
        >
          <Switch
            checked={true}
            disabled={false}
            helpText={null}
            id="mediaPttCheckbox"
            label="Push to talk"
            onChange={[Function]}
          />
        </div>
      </React.Fragment>
      <div
        className="modal__SubText"
      >
        Inappropriate broadcasts will be terminated as per the
         
        <a
          href="/terms"
          rel="noopener noreferrer"
          target="_blank"
        >
          terms and conditions
        </a>
      </div>
    </React.Fragment>
  </div>
</Modal>
`;
